compiler/z/codegen/OMRTreeEvaluator.cpp

```diff
@@ -14116,6 +14116,195 @@ bool isSettingOp(uint8_t byteValue, TR::InstOpCode::Mnemonic SI_opcode)
+      generateRILInstruction(cg, TR::InstOpCode::LGFI, node, resultReg, 0, NULL);
+      TR::MemoryReference *vectorSourceMR = generateS390MemoryReference(resultReg, 0, cg, NULL);
+      generateVRSbInstruction(cg, TR::InstOpCode::VLVG, node, VectorSource, valueReg, vectorSourceMR, 3); //R4 is the valueReg
+
+      // load the array of indices into a vector register
+      TR::MemoryReference *vectorIndicesMR = generateS390MemoryReference(addrReg, 0, cg, NULL);
+      generateVSIInstruction(cg, TR::InstOpCode::VLRL, node, VectorIndices, vectorIndicesMR, 15); //R5 is the addrReg
+
+      /// Since we are on BigEndian architecture, bit positions in a register are numbered left to right. Hence
+      /// the most significant byte in the array is loaded into the right most byte (i.e least significant byte in the register)
+      /// The VPERM instruction below uses a mask to reverse the order of the array so we can do the computation correctly.
+      TR::MemoryReference *maskAddrMR = generateS390MemoryReference(cg->findOrCreateConstant(node, mask, 16), cg, 0, node);
+      generateVSIInstruction(cg, TR::InstOpCode::VLRL, node, tmpVector, maskAddrMR, 15);
+      generateVRReInstruction(cg, TR::InstOpCode::VPERM, node, VectorIndices, VectorIndices, VectorIndices, tmpVector, 0, 0);
+
+      /// Registers on the Z hardware have their bit and byte positions numbered from left to right in registers. Thus, a bit position
+      /// of 2 specified by the bitPermute IL is equivalent to 63 - 2 = 61 on Z. Thus the VS instruction below does a "packed subtraction" to
+      /// subtract all bit indices by 63.
+      TR::MemoryReference *inverseAddrMR = generateS390MemoryReference(cg->findOrCreateConstant(node, inverse, 16), cg, 0, node);
+      generateVSIInstruction(cg, TR::InstOpCode::VLRL, node, tmpVector, inverseAddrMR, 15);
+      generateVRRcInstruction(cg, TR::InstOpCode::VS, node, VectorIndices, tmpVector, VectorIndices, 0);
+
+      // The VBPERM instruction now maps perfectly to the IL, so do the operation.
+      generateVRRcInstruction(cg, TR::InstOpCode::VBPERM, node, VectorSource, VectorSource, VectorIndices, 0);
+
+      generateRILInstruction(cg, TR::InstOpCode::LGFI, node, resultReg, 3, NULL);
+
+      // load result back into GPR
+      TR::MemoryReference *loadBackToGPRMR = generateS390MemoryReference(resultReg, 0, cg, NULL);
+      generateVRScInstruction(cg, TR::InstOpCode::VLGV, node, resultReg, VectorSource, loadBackToGPRMR, 1);
+
+      // The result of the VBPERM op is now in resultReg's 48-63 bit locations. Extract the bits you want via a mask
+      // and store the final result in resultReg. This is necessary because VBPERM operates on 16 bit positions at a time.
+      // If the array specified by the bitPermute IL was smaller than 16, then invalid bits can be selected. 
+      int resultMask = 1;
+      for (int j = 1; j < arrayLen; j++)
+         {
+         resultMask = ( (resultMask << 1) | 0x1);
+         }
```

Comment:
Why not replace the whole loop with `resultMask = (1 << arrayLen) - 1;` - double check this is correct.

Reply:
Addressed here: 35f30572a682f47bfb719175aeaa26e877931fb1